drivers: ethernet: Deprecate eth_mcux, and some improvements to nxp_enet

decsny opened the pull request:
See commit messages for descriptions. 

Note that this PR contains a commit cherry-picked from #69207. This is because the ksz8081 phy driver improvements are necessary to unmark the new driver as experimental, because this PHY is on more or less all of the NXP EVKs - and some of the changes are crucial functional fixes.

stanislav-poboril committed:
drivers: ethernet: phy: KSZ8081 PHY Driver improvements

Added changes required for nxp_enet ethernet driver to work
with multiple PHYs and fixed few problems:

- The cfg_link API resets PHY before configuring link. It was moved
  here so the ethernet driver does not have to reset it - not all
  PHYs need reset before configuring link and moving the reset code
  here makes possible to have the reset done in a PHY specific way
  (for example to reset by toggling GPIO pin). It also avoids ethernet
  driver touching PHY registers without locking.
- When reset GPIO is not defined, reset is performed by setting reset
  bit in control register.
- The cfg_link API does not return error when autonegotiation fails.
  This fixes situation when the link is down at system start - ethernet
  driver then skipped setting link-change callback and link was never
  to be detected again.
- Added reset of excessive bits 16-31 when reading register values.
  As only 16 bits are read from PHY, but the API is supposed to read
  into uint32_t, the remaining bits contained previous data after
  a successful read.
- Fixed missing mutex unlock when querying link state and link was down.
- Added missing initializer to link state variables. This could result
  in link state change detection while link was still down, because
  the speed/duplex settings could be random and old and new state
  could be wrongly detected as different.
- Not logging link speed/duplex status when link is not up.

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

Commit:
drivers: nxp_enet: Correct PTP clock dependencies

The dependencies should be in a 'depends on' clause.
Also, 'depends on PTP_CLOCK' is redundant because this is
within 'if PTP_CLOCK' already.

Additionally, conditionally include the PTP header in the mac driver.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: nxp_enet: Remove unnecessary comments

Some comments in this driver are just redundant - the code
spells out exactly what the comment says. And some comment
blocks are just unnecessary to have. Finally, remove the TODO
comment because this is flagged by static analysis and there
are already tracking issues for the zero copy enhancement.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: nxp_enet: Remove unused interrupt codes

Remove code that does nothing, regarding unused interrupts.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: nxp_enet: Add NET DEVICE PM function

Add functionality for when NET_POWER_MANAGEMENT (ie PM_DEVICE)
is enabled. This function code originally comes from the
old eth_mcux driver which was only tested on kinetis family SOCs,
so that family kconfig is a dependency for this feature for now.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: nxp_enet: Convert RX to workqueue

Convert from multiple threads for each instance to use
one workqueue for all instances. The benefit is to save
memory and use a kernel function that already exists
for a use case like this.

Also introduce the ETH_NXP_ENET_RX_THREAD_PRIORITY kconfig,
which makes the thread priority of the workqueue configurable.

Finally, remove the code enabling the RxBufferInterrupt, since
the meaning of it isn't used currently in this driver.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: nxp_enet: Generate MAC using eth.h

The MAC address macros are ridiculous in this driver.
Rewrite to be simpler and use eth.h common function.
Also, clarify the mac address generation on the DT overlays.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: nxp_enet: Deprecate ETH_MCUX

Deprecate ETH_MCUX, by:

- Marking it as DEPRECATED in Kconfig, obviously.
- Unmarking the new driver as experimental.
- Putting the new and old drivers in the same folder.
- Reworking the menu appearance of the driver selection.
  Note that technically now it is possible to choose the wrong
  driver than what is enabled in DT, this is intentional, but
  the correct one will obviously be enabled by default.
- Convert all sample overlays to the new Kconfigs. This was
  part of the motivation for the shared overlays, as it was
  causing twister/CI logs to look ugly and misleading due
  to Kconfig warnings.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
dts: nxp: Convert ENET DT default to new binding.

Convert all of the NXP SOCs with ENET to use the new
binding scheme, which is used by the new driver.

Convert any boards using this SOC to the new scheme as well,
and remove from the documentation the bit about the experimental
nature of the new driver and the overlay that shall no longer exist.

Some of the boards I do not have the hardware of, so apologies
if something breaks, as I have no way to know. All the boards
were made sure to at least build.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

danieldegrasse commented on drivers/ethernet/nxp_enet/Kconfig:
```diff
@@ -1,34 +1,91 @@
 	select NOCACHE_MEMORY if HAS_MCUX_CACHE && CPU_HAS_DCACHE
 	select ARM_MPU if CPU_CORTEX_M7
 	select NET_POWER_MANAGEMENT if PM_DEVICE
+	select DEPRECATED
```
We should add something to the release notes highlighting this change

dleach02 replied:
agreed. Need to add to the release notes and probably the migration guide (maybe)

decsny replied:
Yes I was thinking about this I just haven't got to it yet (PR has a few loose ends I need to address in the morning)

danieldegrasse commented on drivers/ethernet/nxp_enet/eth_mcux.c:
I get the reasoning behind renaming the driver and causing a build error for users, but I think the point of `CONFIG_DEPRECATED` is to help users transition. IMO we should keep Kconfigs in the old driver as is

decsny replied:
Okay, I knew this would be controversial and I don't really care that much so I will just not rename them. 

Although I have to state an observation:

Small build errors from successive versions of zephyr are purposefully introduced all the time. The only reason we seem to want to make this depreciation a seamless transition is because the change is somewhat bigger, and therefore scarier. There appears to be an counterintuitive *inverse relationship* between the magnitude of a change being made and the amount of disruption produced by that change that can be tolerated. 

An opinion about this. Part of this is because we *decided* to do this this way in this case. It's actually very common for us to rewrite drivers and completely change DT binding schema and rename kconfigs. All the time in fact... to name a few: display drivers, LPADC, the PIT counter... but with this we decided it is different, even though it is not really. The real truth here (on the overall introduction of a second driver) is that the only obvious difference from a user perspective is configuration options, just like any other change we regularly make.

So my only point here is that I think that should be consistent in how we present changes to users. Whether we do it with this deprecation strategy (for which this has been a good experiment), or the way we normally do it, to me it doesn't matter, clearly user feedback would be useful, but right now it seems the deciding factor on a course of action for the process of change is the magnitude of fear of the change, which, in my opinion, does not facilitate a valid strategy.

dleach02 replied:
Deprecation is a guiding principle of Zephyr project. 

Step 1 (release N): Add an experimental feature to replace an existing feature
Step 2 (release N+1): Transition experimental feature to non-experimental and mark legacy as deprecated
Step 3 (release N+3): Retire legacy feature. Note that the deprecation rules state that something stays in deprecation state for two release cycles before retiring.

Each of these steps shall ensure that legacy feature continues  to work to allow the consumer to react on their time and  not ours. 

https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html

decsny replied:
That documentation is for APIs. The closest thing to an API that we are talking about is configuration options (DT and kconfig). Currently everything selecting the CONFIG_DEPRECATED symbol is either an actual API or some feature that is literally no longer supported. In this case, the feature is the same, the driver is just changing. We could easily have just presented it as we do with any other driver change. We force users to make configuration changes all of the time, such as the examples I listed above and plenty more. If it's the implementation that we're concerned about, then what do we have to say then about the massive HAL updates that we do every few months that change millions of lines of thousands of files, that no single reviewer can even nearly begin to comprehend in full? They aren't API changes to the user, so we don't do a deprecation process of every HAL version, because it's all implementation. How is this different? 

If we want to do it this way from now on for any implementation change or configuration option update, all I'm saying is we should be consistent. Don't understand exactly what is special about this case that it is being treated like an API.

danieldegrasse commented on boards/nxp/mimxrt1160_evk/dts/nxp,enet-experimental.overlay:
We should remove the testcase yaml cases as well now, since they are no longer needed

decsny replied:
Indeed

decsny commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -744,57 +746,6 @@ static const struct ethernet_api api_funcs = {
-#define NXP_ENET_GENERATE_MAC_UNIQUE(n)						\
-	static void generate_eth_##n##_mac(uint8_t *mac_addr)			\
-	{									\
-		uint32_t id = ETH_NXP_ENET_UNIQUE_ID;				\
```
I agree, I deleted it on accident, but, it needs to have a new DT property introduced then because, as discussed on some other open tickets to ethernet drivers, this behavior is not part of the common zephyr/DT API, it either needs to be documented that the default behavior if a node has no Mac address is to use a pretermined one, or there needs to be another explicit DT property to clarify it. Right now a lot of drivers leave this to completely undefined behavior or even explicitly statically assert this case and cause a build error if the mac address is not specified in DT.

So the previous behavior is not standard and cannot be kept as it is... but I will definitely address adding the feature somewhat properly in this PR. Although I will not take any approach that changes the common ethernet interface in DT or documentation regarding this in this PR as that would require ensuring it to be true for all other ethernet drivers which is a scope creep.

decsny replied:
To clarify, what I am going to just do for now in this PR is introduce a boolean option in the nxp,enet binding to accomplish this, in a followup PR I will investigate a standard way to do it since it is a common feature and this keeps repearedly coming up as a point of debate on other ethernet driver PRs as well

decsny replied:
Also, the fact that the old code is setting the LAA bit seems incorrect to me, since what we are literally doing is the description of a UAA address, ie, using the vendor OUI and then setting the rest based on some unique silicon id. The only controversy about is we are deciding this in OS software rather than simply reading the entire address from some ROM. I still think it's not right to set the LAA bit here, but I'm not sure.

pdgendt commented:
Will the new nxp_enet driver work with Distributed Switch Architecture `DSA`?

From `drivers/ethernet/Kconfig.dsa`:
```
menuconfig NET_DSA
	bool "Distributed Switch Architecture support"
	depends on ETH_MCUX || ETH_SAM_GMAC || ETH_STM32_HAL
	help
	  Enable Distributed Switch Architecture support. For now it
	  only supports Kinetics and STM32 ENET drivers.
```

decsny commented:
> Will the new nxp_enet driver work with Distributed Switch Architecture `DSA`?
> 
> From `drivers/ethernet/Kconfig.dsa`:
> 
> ```
> menuconfig NET_DSA
> 	bool "Distributed Switch Architecture support"
> 	depends on ETH_MCUX || ETH_SAM_GMAC || ETH_STM32_HAL
> 	help
> 	  Enable Distributed Switch Architecture support. For now it
> 	  only supports Kinetics and STM32 ENET drivers.
> ```

I should hopefully get the Segger IPk66f board soon to test it.... but it should have the same dsa related code as eth_mcux did

Commit:
drivers: clock_control: mcux_sim: PTP clock

Add PTP clock get rate code

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
doc: migration-guide-3.7: Add NXP ENET info

Add to migration guide info about the nxp,enet binding change.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: ethernet: DSA supported on ETH_NXP_ENET

I have tested the DSA support with ETH_NXP_ENET driver and it
works, so marking it as supported by adding it to the dependency
list for the DSA feature.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@pdgendt I got one of the segger ipk66f boards as I mentioned, and created a little switched network consisting of:
1. The ipk66f board using the new nxp_enet driver and acting as a switch using DSA
2. A linux PC on one of the dsa slave ports
3. frdm k64f running with the new nxp_enet driver on one of the slave ports
4. frdm mcx947 board running with the nxp_enet_qos driver that just merged

did some developer testing using this network to verify that the driver seems to be working for DSA, so added the commit to show support for it

jukkar commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -88,6 +87,24 @@ struct nxp_enet_mac_data {
 	uint8_t *rx_frame_buf;
 };
 
+K_THREAD_STACK_DEFINE(enet_rx_stack, CONFIG_ETH_NXP_ENET_RX_THREAD_STACK_SIZE);
```
Perhaps `static` ?

jukkar commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -387,22 +404,21 @@ static int eth_nxp_enet_rx(const struct device *dev)
+		return;
 	}
+
+	while (eth_nxp_enet_rx(dev) == 1)
+		;
```
missing `{}` although this looks just fine too. I think there was some Misra rule for this, and a compliance checker script might complain about the missing braces.

Unrelated to this PR, I noticed while checking what `eth_nxp_enet_rx()` is doing, that it might make sense in that function to return the real error from `net_recv_data()` instead of always `-EIO`

decsny replied:
WAIT_FOR didn't make sense to me to use because, we aren't waiting on something, ie there is no notion of time here, it's just repeat this routine until done, so I elected to convert it to a do-while with a return variable

jukkar reviewed (commented):
Couple of minor nits, and the CI needs to be fixed, but overall LGTM

pdgendt commented on drivers/ethernet/Kconfig.dsa:
```diff
@@ -6,7 +6,7 @@
 menuconfig NET_DSA
 	bool "Distributed Switch Architecture support"
-	depends on ETH_MCUX || ETH_SAM_GMAC || ETH_STM32_HAL
+	depends on ETH_MCUX || ETH_SAM_GMAC || ETH_STM32_HAL || ETH_NXP_ENET
```
Thanks for this addition, however this doesn't really scale, and doesn't allow for OOT drivers.

Maybe we need to add a hidden kconfig, something like `ETH_SUPPORT_DSA`, which is selected by those drivers?

```
config ETH_DSA_SUPPORT
	bool
	help
	  Hidden symbol to indicate DSA is supported by an Ethernet driver.

menuconfig NET_DSA
	bool "Distributed Switch Architecture support"
	depends on ETH_DSA_SUPPORT
```

decsny replied:
okay, I added a commit converting this

Commit:
drivers: ethernet: Add ETH_DSA_SUPPORT

Make DSA dependent on ETH_DSA_SUPPORT which is selected by driver
kconfigs, rather than having a list of dependencies.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -67,6 +67,36 @@ Device Drivers and Devicetree
+  :dtcompatible:`nxp,enet`. All in tree SOCs were converted to use this new schema.
+  Thus, all boards using NXP's ENET peripheral will need to align to this binding
+  in DT, which also comes with a different version driver. Alternatively,
+  the ethernet node can be deleted and redefined as the old binding to use
```
I took this to mean that you think the word "ethernet" in this line should be capitalized, which I updated

pdgendt commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -88,6 +87,24 @@ struct nxp_enet_mac_data {
+	struct k_work_queue_config cfg = {.name = "ENET_RX"};
+
+	k_work_queue_init(&rx_work_queue);
+	k_work_queue_start(&rx_work_queue, enet_rx_stack,
+				K_THREAD_STACK_SIZEOF(enet_rx_stack),
+				K_PRIO_COOP(CONFIG_ETH_NXP_ENET_RX_THREAD_PRIORITY),
+				&cfg);
```
Nit: align

kartben commented:
Re-assigning to @decsny - although he's the author of this PR, he's the most appopriate assignee in this particular instance
